nilearn/regions/connectivity.py

```diff
@@ -0,0 +1,392 @@
+        3D NIFTI file containing a mask, which restricts the voxels used in
+        the analysis.
+    thresh : str
+        Threshold value, correlation coefficients lower than this value
+        will be removed from the matrix (set to zero).
+
+    Returns
+    -------
+    W : Compressed Sparse Matrix
+        A Scipy sparse matrix, with weights corresponding to the spatial
+        correlation between the time series from voxel i and voxel j
+
+    References
+    ----------
+    .. [1] Craddock, R. C., James, G. A., Holtzheimer, P. E., Hu, X. P., &
+      Mayberg, H. S. (2012). A whole brain fMRI atlas generated via
+      spatially constrained spectral clustering. Human Brain Mapping.
+      https://doi.org/10.1002/hbm.21333
+
+    """
+    from scipy.sparse import csc_matrix
+    from itertools import product
+
+    neighbors = np.array(
+        sorted(
+            sorted(
+                sorted(
+                    [list(x) for x in list(set(product({-1, 0, 1},
+                                                       repeat=3)))],
+                    key=lambda k: (k[0]),
+                ),
+                key=lambda k: (k[1]),
+            ),
+            key=lambda k: (k[2]),
+        )
+    )
+
+    # Read in the mask
+    msz = clust_mask_img.shape
+
+    # Convert the 3D mask array into a 1D vector
+    mskdat = np.reshape(
+        np.asarray(
+            clust_mask_img.dataobj).astype("bool"),
+        prod(msz))
+
+    # Determine the 1D coordinates of the non-zero elements of the mask
+    iv = np.nonzero(mskdat)[0]
+    sz = func_img.shape
+
+    # Reshape fmri data to a num_voxels x num_timepoints array
+    func_data = _safe_get_data(func_img).astype(np.float32)
+    func_data = np.reshape(func_data, (prod(sz[:3]), sz[3]))
+
+    # Mask the datset to only the in-mask voxels
+    func_data = func_data[iv, :]
+    func_data_sz = func_data.shape
+
+    # Z-score fmri time courses, this makes calculation of the
+    # correlation coefficient a simple matrix product
+    func_data_s = np.tile(np.std(func_data, 1), (func_data_sz[1], 1)).T
+
+    # Replace 0 with large number to avoid div by zero
+    func_data_s[func_data_s == 0] = 1000000
+    func_data_m = np.tile(np.mean(func_data, 1), (func_data_sz[1], 1)).T
+    func_data = (func_data - func_data_m) / func_data_s
+
+    # Set values with no variance to zero
+    func_data[func_data_s == 0] = 0
+    func_data[np.isnan(func_data)] = 0
+
+    # Remove voxels with zero variance, do this here so that the mapping will
+    # be consistent across subjects
+    vndx = np.nonzero(np.var(func_data, 1) != 0)[0]
+    iv = iv[vndx]
+    m = len(iv)
+    print(m, " # of non-zero valued or non-zero variance voxels in the mask")
+
+    # Construct a sparse matrix from the mask
+    msk = csc_matrix(
+        (vndx + 1, (iv, np.zeros(m))), shape=(prod(msz), 1), dtype=np.float32
+    )
+
+    sparse_i = []
+    sparse_j = []
+    sparse_w = [[]]
+
+    for i in range(0, m):
```

Comment:
It is not possible to write such a loop on voxels: this would be horribly slow for large arrays. You need to operate directly at the array level. Numpy has enough functions for that.

Reply:
It actually finishes in a few minutes at most on the whole brain and <30 seconds for a typical ROI. That said, I can gladly look into other approaches.  Since my time is very limited at the moment, I probably wouldn't be able to rewrite this entirely right now if that's what you think would be required.